Nodejs/Product/TypeScriptSourceMapReader/SourceMapReader.cs

```diff
@@ -0,0 +1,290 @@
+//    permissions and limitations under the License.
+//
+//*********************************************************//
+
+using Newtonsoft.Json;
+using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace Microsoft.NodejsTools.TypeScriptSourceMapReader {
+    /// <summary>
+    /// Helper class to read the source map and get the decoded source map
+    /// It also has some apis to test if the url is js url and get the sourceMap url
+    /// </summary>
+    public class SourceMapReader : SourceMapTextReader {
+        #region ApiLoadSourceMapping
+        /// <summary>
+        /// Gets the decoded source map for the given script document
+        /// </summary>
+        /// <param name="scriptFilePathOrUrl">FilePath or Url of the js script file</param>
+        /// <param name="sourceMapUrl">url of the source map</param>
+        /// <returns>DecodedSourceMap if the mapping present otherwise null, throws SourceMapReaderException if read fails</returns>
+        public DecodedSourceMap LoadSourceMap(string scriptFilePathOrUrl, string sourceMapUrl) {
+            // Read source map
+            Uri sourceMapUri;
+            var encodedSourceMap = this.ReadEncodedSourceMap(scriptFilePathOrUrl, sourceMapUrl, out sourceMapUri);
+            return this.DecodeSourceMap(encodedSourceMap, scriptFilePathOrUrl, sourceMapUri);
+        }
+
+        /// <summary>
+        /// Gets the decoded source map for the given script document's source map data
+        /// </summary>
+        /// <param name="sourceMapText">source map text - to decode</param>
+        /// <param name="scriptFilePathOrUrl">FilePath or Url of the js script file</param>
+        /// <param name="sourceMapUri">Uri of the source map (that was used to read the source map)</param>
+        /// <returns>DecodedSourceMap if the mapping present otherwise null, throws SourceMapReaderException if read fails</returns>
+        public DecodedSourceMap LoadSourceMapFromSourceMapText(string sourceMapText, string scriptFilePathOrUrl, Uri sourceMapUri) {
+            // Convert the sourceMapText into EncodedSourceMap json format
+            EncodedSourceMap encodedSourceMap;
+            if (sourceMapText == null) {
+                // Read the source Map from sourceMapUri
+                encodedSourceMap = this.ReadEncodedSourceMap(sourceMapUri);
+            } else {
+                // Read source map using sourceMapText
+                StringReader stringReader = null;
+                try {
+                    stringReader = new StringReader(sourceMapText);
+                    encodedSourceMap = this.ReadEncodedSourceMap(stringReader);
+                } catch (Exception e) {
+                    throw new SourceMapReadFailedException(sourceMapUri.IsFile ? sourceMapUri.LocalPath : sourceMapUri.AbsoluteUri, e);
+                } finally {
+                    if (stringReader != null) {
+                        stringReader.Close();
+                    }
```

Comment:
Instead of a `finally`, put a  `StringReader` `using` statement inside of the `try`.